text/0101-revamping-the-sdk-performance-api.md

```diff
@@ -0,0 +1,268 @@
+
+Summarizing, here are the core Issues in SDK Performance API:
+
+1. Users have to understand the difference between spans/transactions and their schema differences.
+2. Users have to set/get transactions/spans from a scope (meaning they also have to understand what a scope/hub means).
+3. Nesting transactions within each other is undefined behaviour - no obvious way to make a transaction a child of another.
+4. If a transaction finishes before it’s child span finishes, that child span gets orphaned and the data is never sent to Sentry. This is most apparent if you have transactions that automatically finish (like on browser/mobile).
+5. Transactions have a max child span count of 1000 which means that eventually data is lost if you keep adding child spans to a transaction.
+
+# Improvements
+
+## New SDK API
+
+The new SDK API has the following requirements:
+
+1. Newly created spans must have the correct trace and parent/child relationship
+2. Users shouldn’t be burdened with knowing if something is a span/transaction
+3. Spans only need a name to identify themselves, everything else is optional.
+4. The new top level APIs should be as similar to the OpenTelemetry SDK public API as possible.
+
+There are two top level methods we'll be introducing to achieve this: `Sentry.startActiveSpan` and `Sentry.startSpan`. `Sentry.startActiveSpan` will take a callback and start/stop a span automatically. In addition, it'll also set the span on the current scope. Under the hood, the SDK will create a transaction or span based on if there is already an existing span on the scope.
+
+```ts
+namespace Sentry {
+  declare function startActiveSpan<T>(
+    spanContext: SpanContext,
+    callback: (span: Span) => T
+  ): T;
+}
+
+// span that is created is provided to callback in case additional
+// attributes have to be added.
+// ideally callback can be async/sync
+const returnVal = Sentry.startActiveSpan(
+  { name: "expensiveCalc" },
+  (span: Span) => expensiveCalc()
+);
+
+// If the SDK needs async/sync typed different we can expose this
+// declare function startActiveSpanAsync<T>(
+//   spanContext: SpanContext,
+//   callback: (span: Span) => Promise<T>,
+// ): Promise<T>;
+```
+
+In the ideal case, `startActiveSpan` should generally follow this code path.
+
+1. Get the active span from the current scope
+2. If the active span is defined, create a child span of that active span based on the provided `spanContext`, otherwise create a transaction based on the provided `spanContext`.
+3. Run the provided callback
+4. Finish the child span/transaction created in step 2
+5. Remove the child span/transaction from the current scope and if it exists, set the previous active span as the active span in the current scope.
```

Comment:
I'm wondering if this solves (or not) the use case that you have an active span (span A), and spawn multiple threads to download something, and each of those threads will start a new span (span B and C), but then the relationship of the spans are messed up because they "race each other", for example, span B and C should be a child of span A, but what I see is that span C is a child of span B.

Reply:
I guess the caveat here is that the hub/scope needs to be properly forked and correctly follow the execution context.

I guess in single scope instances, this logic breaks down re: parent-child relationships. Let me amend the RFC to reference this.